src/core/src/service_interfaces/TelemetryWriter.py

```diff
@@ -142,3 +112,120 @@ def get_disk_type(self):
         else:
             return "Unknown"
     # end region
+
+    def __new_event_json(self, event_level, message, task_name):
+        return {
+            "Version": Constants.EXT_VERSION,
+            "Timestamp": str(datetime.datetime.utcnow()),
+            "TaskName": task_name,
+            "EventLevel": event_level,
+            "Message": self.__ensure_message_restriction_compliance(message),
+            "EventPid": "",
+            "EventTid": "",
+            "OperationId": self.__operation_id  # activity id from from config settings
+        }
+
+    def __ensure_message_restriction_compliance(self, full_message):
+        """ Removes line breaks, tabs and restricts message to a byte limit """
+        message_size_limit_in_bytes = Constants.TELEMETRY_MSG_SIZE_LIMIT_IN_BYTES
+        formatted_message = re.sub(r"\s+", " ", str(full_message))
+
+        if len(formatted_message.encode('utf-8')) > message_size_limit_in_bytes:
+            self.composite_logger.log_telemetry_module("Data sent to telemetry will be truncated as it exceeds size limit. [Message={0}]".format(str(formatted_message)))
+            formatted_message = formatted_message.encode('utf-8')
+            bytes_dropped = len(formatted_message) - message_size_limit_in_bytes + Constants.TELEMETRY_BUFFER_FOR_DROPPED_COUNT_MSG_IN_BYTES
+            return formatted_message[:message_size_limit_in_bytes - Constants.TELEMETRY_BUFFER_FOR_DROPPED_COUNT_MSG_IN_BYTES].decode('utf-8') + '. [{0} bytes dropped]'.format(bytes_dropped)
+
+        return formatted_message
+
+    def write_event(self, message, event_level=Constants.TelemetryEventLevel.Informational, task_name=Constants.TELEMETRY_TASK_NAME):
+        """ Creates and writes event to event file after validating none of the telemetry size restrictions are breached """
+        # if self.events_folder_path is None or not os.path.exists(self.events_folder_path) or not Constants.TELEMETRY_ENABLED_AT_EXTENSION:
+        if not self.__is_agent_compatible or not Constants.TELEMETRY_ENABLED_AT_EXTENSION:
+            return
+
+        self.__delete_older_events()
+
+        event = self.__new_event_json(event_level, message, task_name)
+        if len(json.dumps(event)) > Constants.TELEMETRY_EVENT_SIZE_LIMIT_IN_BYTES:
+            self.composite_logger.log_telemetry_module_error("Cannot send data to telemetry as it exceeded the acceptable data size. [Data not sent={0}]".format(json.dumps(message)))
+        else:
+            self.write_event_using_temp_file(self.events_folder_path, event)
```

Comment:
Generic comment about telemetry_writer -- the only error from this that may show up in customer facing statuses, etc. should be something generic like 'Internal reporting error. Execution could not complete.' which is piped up the call stack with no other details (thus the technical requirement is a try catch that accomplishes that while logging the specific details within the log file.

We should never have a customer report that they saw such a message since it'd 99.99% be a bug in our code.

Reply:
This may require try catches on most/all externally invokable methods.